### PR TITLE
PWN-8779 - Add Striga banners 

### PR DESCRIPTION
**Pull request by Davran331122**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-8779

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 1]
```diff
@@ -313,6 +315,15 @@ class HomePresenter(
         }
     }
 
+    override fun onBannerClicked(bannerTitleId: Int) {
+        val statusFromKycBanner = strigaUiBannerMapper.onBannerClicked(bannerTitleId)
+        view?.showStrigaKyc(statusFromKycBanner)
+    }
+
+    override fun onBannerCloseClicked(bannerTitleId: Int) {
+        TODO("Not yet implemented")
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:324]: ?

File: app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            while (isActive) {
+                try {
+                    val kycStatusDestination = loadUserStatus().let(::mapToDestination)
+                    strigaUserDestinationFlow.value = kycStatusDestination
+                    delay(DELAY_IM_MILLISECONDS)
+                } catch (e: Throwable) {
+                    Timber.tag(TAG).e(e, "Error while fetching user kyc status")
+                }
+            }
+        }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt:50]: I can't see where you are cancelling this coroutine in case of minimizing the app. 
Or you can use `AppScope` here

**Review by 1sakov — APPROVED** [review]
lgtm, minor comments

File: app/src/main/java/org/p2p/wallet/kyc/model/StrigaKycSignUpStatus.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
+package org.p2p.wallet.kyc.model
+
+import androidx.annotation.ColorRes
+import androidx.annotation.DrawableRes
+import androidx.annotation.StringRes
+import org.p2p.wallet.R
+
+enum class StrigaKycSignUpStatus(
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/kyc/model/StrigaKycSignUpStatus.kt]: It's not clear that the model used for banners, consider to rename it to something like:
- StrigaKycBannerStatus
- StrigaKycSignupBannerStatus
- StrigaKycSignupBannerData
- other options with *banner*

Btw, we use `Signup` instead of `SignUp`

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 4]
```diff
@@ -341,16 +352,6 @@ class HomePresenter(
         }
     }
 
-    private fun loadEthTokens() {
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:344]: Why did you remove it?

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 5]
```diff
@@ -479,14 +480,8 @@ class HomePresenter(
...
 
-        val homeBannerItem = HomeBannerItem(
-            id = R.id.home_banner_top_up,
-            titleTextId = R.string.main_banner_title,
-            subtitleTextId = R.string.main_banner_subtitle,
-            buttonTextId = R.string.main_banner_button,
-            drawableRes = R.drawable.ic_main_banner,
-            backgroundColorRes = R.color.bannerBackgroundColor
-        )
+        val homeBannerItem = strigaUiBannerMapper.mapToBigBanner(StrigaKycStatusBanner.VERIFICATION_DONE)
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:483]: is it a debugging code?

File: app/src/main/java/org/p2p/wallet/kyc/StrigaFragmentFactory.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import org.p2p.wallet.kyc.model.StrigaKycStatusBanner
+import org.p2p.wallet.smsinput.SmsInputFactory
+import org.p2p.wallet.striga.finish.StrigaSignupFinishFragment
+import org.p2p.wallet.striga.onboarding.StrigaOnboardingFragment
+import org.p2p.wallet.striga.signup.ui.StrigaSignUpFirstStepFragment
+import org.p2p.wallet.striga.user.model.StrigaUserStatusDestination
+
+class StrigaFragmentFactory() {
+
+    fun kycFragment(status: StrigaKycStatusBanner): Fragment {
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/kyc/StrigaFragmentFactory.kt:13]: As I know what you need is just return `StrigaKycFragment()` (our wrapper) for any status, it seems sumsub handles all this stuff by itself
@gslevinkov please, elaborate on this

File: app/src/main/java/org/p2p/wallet/striga/ui/TopUpWalletPresenter.kt [hunk 7]
```diff
@@ -11,29 +9,20 @@ import org.p2p.wallet.common.feature_toggles.toggles.remote.StrigaSignupEnabledF
...
     private val appFeatureFlags: InAppFeatureFlags,
     private val userInteractor: UserInteractor,
     private val strigaSignupFeatureToggle: StrigaSignupEnabledFeatureToggle,
     private val seedPhraseProvider: SeedPhraseProvider,
-    private val strigaUserInteractor: StrigaUserInteractor,
+    private val strigaUserStatusRepository: StrigaUserStatusRepository
 ) : BasePresenter<TopUpWalletContract.View>(),
     TopUpWalletContract.Presenter {
 
     private val strigaBankTransferProgress = MutableSharedFlow<Boolean>(replay = 1)
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/striga/ui/TopUpWalletPresenter.kt:25]: Please pay attention when merging with the develop branch to metadata loading. We must ensure that the user cannot proceed until the metadata has fully loaded

File: app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt (hunk 2 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt:57]: We decided to make single request

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt (hunk 4 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:344]: it's was not used

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/SingleBannerViewHolder.kt [hunk 8]
```diff
@@ -21,7 +22,7 @@ class SingleBannerViewHolder(
         parent: ViewGroup,
         listener: OnHomeItemsClickListener
     ) : this(
-        binding = ItemBannerSingleBinding.inflate(LayoutInflater.from(parent.context), parent, false),
+        binding = ItemHomeBannerBinding.inflate(LayoutInflater.from(parent.context), parent, false),
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/adapter/SingleBannerViewHolder.kt:25]: do binding like we do now
```
SingleBannerViewHolder(
        parent: ViewGroup,
        binding: ItemHomeBannerBinding = parent.inflateViewBinding(attachToParent = false),
        listener: OnHomeItemsClickListener
    )
```
without secondary constructor

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/SingleBannerViewHolder.kt [hunk 9]
```diff
@@ -33,21 +34,32 @@ class SingleBannerViewHolder(
...
+
+    private fun onBind(item: StrigaBanner) = with(binding) {
+        val item = item.status
+        textViewTitle.setText(item.bannerTitleResId)
+
+        val subtitleText = context.getString(item.bannerMessageResId)
+        textViewSubtitle.text = subtitleText
+        textViewSubtitle.isVisible = subtitleText.isNotEmpty()
+
+        textViewSubtitle.setText(item.bannerMessageResId)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/adapter/SingleBannerViewHolder.kt:51]: you set text two times

File: app/src/main/java/org/p2p/wallet/home/ui/main/adapter/SingleBannerViewHolder.kt [hunk 10]
```diff
@@ -33,21 +34,32 @@ class SingleBannerViewHolder(
...
-
-            root.setOnClickListener { listener.onBannerClicked(item.optionTextId) }
+    fun onBind(item: HomeScreenBanner) {
+        when (item) {
+            is StrigaBanner -> onBind(item)
+        }
+    }
+
+    private fun onBind(item: StrigaBanner) = with(binding) {
+        val item = item.status
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/adapter/SingleBannerViewHolder.kt:44]: item = item.status? 😬

File: app/src/main/java/org/p2p/wallet/kyc/StrigaFragmentFactory.kt (hunk 6 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/kyc/StrigaFragmentFactory.kt:14]: yep, Striga fetches status by themself and show needed screens

File: app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers
+import org.p2p.wallet.kyc.model.StrigaKycStatusBanner
+import org.p2p.wallet.striga.StrigaUserIdProvider
+import org.p2p.wallet.striga.user.model.StrigaUserStatus
+import org.p2p.wallet.striga.user.model.StrigaUserStatusDestination
+
+private const val DELAY_IM_MILLISECONDS = 10_000L
+private const val TAG = "StrigaUserStatusRepository"
+
+class StrigaUserStatusRepository(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt:18]: i don't see any interface

File: app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    private val userRepository: StrigaUserRepository,
+) : CoroutineScope {
+
+    override val coroutineContext: CoroutineContext
+        get() = dispatchers.io
+
+    private val strigaUserDestinationFlow = MutableStateFlow(StrigaUserStatusDestination.NONE)
+    private val strigaBannerFlow = MutableStateFlow<StrigaKycStatusBanner?>(null)
+
+    fun getUserDestinationFlow(): Flow<StrigaUserStatusDestination> = strigaUserDestinationFlow
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/striga/user/repository/StrigaUserStatusRepository.kt:31]: getUserDestinationFlow -> observerUserDestination
getBannerFlow -> observerKycBanner

**Review by gslevinkov — APPROVED** [review]
